hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -574,52 +575,60 @@ public void dispatchEvent(Object event, CacheEventListener listener) {
+                                                                 EventFilter eventFilter, boolean isLocal) {
         return registerListenerInternal(cacheNameWithPrefix, listener, eventFilter, isLocal);
     }
 
-    protected UUID registerListenerInternal(String cacheNameWithPrefix, CacheEventListener listener,
-                                            EventFilter eventFilter, boolean isLocal) {
+    protected CompletableFuture<EventRegistration> registerListenerInternal(String cacheNameWithPrefix,
+                                                                            CacheEventListener listener, EventFilter eventFilter,
+                                                                            boolean isLocal) {
         EventService eventService = getNodeEngine().getEventService();
-        EventRegistration reg;
+        CompletableFuture<EventRegistration> registrationFuture;
         if (isLocal) {
             if (eventFilter == null) {
-                reg = eventService.registerLocalListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, listener);
+                registrationFuture = eventService
+                        .registerLocalListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, listener);
             } else {
-                reg = eventService.registerLocalListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix,
-                        eventFilter, listener);
+                registrationFuture = eventService
+                        .registerLocalListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, eventFilter, listener);
             }
         } else {
             if (eventFilter == null) {
-                reg = eventService.registerListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, listener);
+                registrationFuture = eventService
+                        .registerListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, listener);
             } else {
-                reg = eventService.registerListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix,
-                        eventFilter, listener);
+                registrationFuture = eventService
+                        .registerListener(AbstractCacheService.SERVICE_NAME, cacheNameWithPrefix, eventFilter, listener);
             }
         }
 
-        UUID id = reg.getId();
-        if (listener instanceof Closeable) {
-            closeableListeners.put(id, (Closeable) listener);
-        } else if (listener instanceof CacheEntryListenerProvider) {
-            CacheEntryListener cacheEntryListener = ((CacheEntryListenerProvider) listener).getCacheEntryListener();
-            if (cacheEntryListener instanceof Closeable) {
-                closeableListeners.put(id, (Closeable) cacheEntryListener);
+        registrationFuture.whenCompleteAsync((BiConsumerEx<EventRegistration, Throwable>) (eventRegistration, throwable) -> {
```

Comment:
Cause of the failure https://github.com/hazelcast/hazelcast/issues/15935
We should return the returned complationStage from `whenCompleteAsync` method.